# **Achievement No. 6: Advanced Analytics and Dashboard Design (I)**

## 2026 QS World University Rankings

#### **Summary**

##### QS stands for Quacquarelli Symonds. It is a provider of services, analytics, and insight to the global higher education sector.

##### The QS World University Rankings is therefore QS's ranking of world's top universities. This ranking seems to be widely recognized and influential.

#### **Contents**

##### 1) Importing Libraries

##### 2) Importing Datasets

##### 3) Accuracy and Consistency Checks

##### 4) Completeness (Missing Values)

##### 5) Addressing Missing Values

##### - Addressing Missing Values in the 'Previous Rank', 'Size', 'Research', and 'Status' columns

##### - Addressing Missing Values in the Ranks and Scores columns

#### **1) Importing Libraries**

In [1]:
import pandas as pd
import numpy as np
import os

#### **2) Importing Datasets**

In [2]:
# Setting up the path

path = r'C:\Users\andd0\Documents\Jupyter_Advanced Analytics and Dashboard Design'

In [3]:
# Importing data sets

df_2026_ranking = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', '2026 QS World University Rankings CSV.csv'), index_col = False)

In [4]:
# Note: I may need to use the 2025 ranking. I'm importing it here, too. Due to an error message, I had to use the 'lating1' encoding

file_path = os.path.join(path, '02 Data', 'Original Data', '2025 QS World University Rankings CSV.csv')

df_2025_ranking = pd.read_csv(file_path, index_col=False, encoding='latin1')

##### **Checking features of the 2026_ranking dataset**

In [5]:
# Columns and rows

df_2026_ranking.shape

(1501, 30)

In [6]:
# Data types

df_2026_ranking.dtypes

2026 Rank             object
Previous Rank         object
Institution Name      object
Country/Territory     object
Region                object
Size                  object
Focus                 object
Research              object
Status                object
AR SCORE             float64
AR RANK               object
ER SCORE             float64
ER RANK               object
FSR SCORE            float64
FSR RANK              object
CPF SCORE            float64
CPF RANK              object
IFR SCORE            float64
IFR RANK              object
ISR SCORE            float64
ISR RANK              object
ISD SCORE            float64
ISD RANK              object
IRN SCORE            float64
IRN RANK              object
EO SCORE             float64
EO RANK               object
SUS SCORE            float64
SUS RANK              object
Overall SCORE         object
dtype: object

##### Having a quick look at the 2025 ranking dataset

In [7]:
df_2025_ranking.shape

(1503, 28)

#### **3) Accuracy and Consistency Checks**

##### The data types are OK for the purposes of this analysis, except by the data type of the column *Overal SCORE*.

##### I may want to make some calculations for that column; therefore, I'll change its data type to number.

In [8]:
# This column contains hyphens; therefore, I first need to change those to NaN before I can change the column data type.

# Replace '-' with NaN
df_2026_ranking['Overall SCORE'] = df_2026_ranking['Overall SCORE'].replace('-', np.nan)

In [9]:
# Convert to float
df_2026_ranking['Overall SCORE'] = df_2026_ranking['Overall SCORE'].astype('float64')

In [10]:
# Checking whether the change was successful (it was)

df_2026_ranking.dtypes

2026 Rank             object
Previous Rank         object
Institution Name      object
Country/Territory     object
Region                object
Size                  object
Focus                 object
Research              object
Status                object
AR SCORE             float64
AR RANK               object
ER SCORE             float64
ER RANK               object
FSR SCORE            float64
FSR RANK              object
CPF SCORE            float64
CPF RANK              object
IFR SCORE            float64
IFR RANK              object
ISR SCORE            float64
ISR RANK              object
ISD SCORE            float64
ISD RANK              object
IRN SCORE            float64
IRN RANK              object
EO SCORE             float64
EO RANK               object
SUS SCORE            float64
SUS RANK              object
Overall SCORE        float64
dtype: object

In [11]:
# Descriptive statistics of quantitative variables

df_2026_ranking.describe()

AR SCORE     ER SCORE    FSR SCORE    CPF SCORE    IFR SCORE  \
count  1501.000000  1501.000000  1501.000000  1501.000000  1414.000000   
mean     25.785943    26.944237    33.950433    30.425516    36.305658   
std      24.500905    25.504494    28.440071    29.679882    35.252024   
min       1.000000     1.000000     1.000000     1.000000     1.000000   
25%       8.800000     8.500000    10.800000     6.000000     6.600000   
50%      16.000000    16.500000    23.500000    18.000000    20.100000   
75%      32.700000    37.500000    50.500000    49.700000    66.100000   
max     100.000000   100.000000   100.000000   100.000000   100.000000   

        ISR SCORE    ISD SCORE    IRN SCORE     EO SCORE    SUS SCORE  \
count  1464.00000  1464.000000  1499.000000  1501.000000  1477.000000   
mean     33.32541    34.526981    53.356905    29.989674    51.254367   
std      32.75066    31.108600    28.920632    29.197573    21.266331   
min       1.00000     1.000000     1.000000     1.000000     3.000000   
25%       5.90000     8.675000    27.500000     6.200000    35.700000   
50%      19.40000    21.700000    55.800000    17.900000    48.700000   
75%      56.70000    55.650000    78.500000    46.000000    66.500000   
max     100.00000   100.000000   100.000000   100.000000   100.000000   

       Overall SCORE  
count     703.000000  
mean       46.756046  
std        18.842125  
min        25.100000  
25%        30.900000  
50%        41.500000  
75%        58.900000  
max       100.000000

In [12]:
# Mode: To prevent Python from giving me a multimode answer (very likely), I specified just the first mode per column

df_2026_ranking.mode().iloc[0]

2026 Rank                           1201-1400
Previous Rank                       1001-1200
Institution Name               ADA University
Country/Territory    United States of America
Region                                   Asia
Size                                        L
Focus                                      FC
Research                                   VH
Status                                 Public
AR SCORE                                  6.4
AR RANK                                  701+
ER SCORE                                  8.2
ER RANK                                  701+
FSR SCORE                               100.0
FSR RANK                                 801+
CPF SCORE                                 1.4
CPF RANK                                 801+
IFR SCORE                               100.0
IFR RANK                                 801+
ISR SCORE                               100.0
ISR RANK                                 801+
ISD SCORE                         

#### **4) Completeness (Missing Values)**

In [13]:
# Checking how many missing values each column has

df_2026_ranking.isnull().sum()

2026 Rank              0
Previous Rank        112
Institution Name       0
Country/Territory      0
Region                 0
Size                   1
Focus                  0
Research               1
Status                47
AR SCORE               0
AR RANK                0
ER SCORE               0
ER RANK                0
FSR SCORE              0
FSR RANK               0
CPF SCORE              0
CPF RANK               0
IFR SCORE             87
IFR RANK              87
ISR SCORE             37
ISR RANK              37
ISD SCORE             37
ISD RANK              37
IRN SCORE              2
IRN RANK               2
EO SCORE               0
EO RANK                0
SUS SCORE             24
SUS RANK              24
Overall SCORE        798
dtype: int64

#### **5) Addressing Missing Values**

##### **Column** *Previous Rank*

##### **Decision:** I will impute the 112 empty values with 'Not Ranked'.

##### **Justification:** After comparing the 2026 and 2025 rankings (xls files), I discovered that these universities do not appear in the 2025 ranking. 

In [14]:
df_2026_ranking['Previous Rank'].fillna('Not Ranked', inplace=True)

C:\Users\andd0\AppData\Local\Temp\ipykernel_23700\245397026.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df_2026_ranking['Previous Rank'].fillna('Not Ranked', inplace=True)


In [15]:
# Checking result (it worked)

df_2026_ranking['Previous Rank'].tail(15)

1486    Not Ranked
1487    Not Ranked
1488    Not Ranked
1489    Not Ranked
1490     1201-1400
1491    Not Ranked
1492         1401+
1493     1201-1400
1494     1201-1400
1495    Not Ranked
1496     1201-1400
1497     1201-1400
1498    Not Ranked
1499     1201-1400
1500     1201-1400
Name: Previous Rank, dtype: object

##### **Columns** *Size & Research*

##### **Decision:** Both columns have 1 missing value each, and they are linked to the same university: Islamic Azad University. I will impute *Size* with XL and Research with 'N/A'.

##### **Justification:** I found the [support article](https://support.qs.com/hc/en-gb/articles/360021876820-QS-Institution-Classifications?utm_source=chatgpt.com) of the QS website which states that universities with more than 30,000 students is considered XL. According to the data available in the [QS website](https://www.topuniversities.com/universities/islamic-azad-university#p2-university-information), the Islamic Azad University has more than 1 million students; therefore, its size is XL. There's no info on its research intensity.

In [16]:
row = df_2026_ranking[df_2026_ranking['Institution Name'] == 'Islamic Azad University']

In [17]:
row

2026 Rank Previous Rank         Institution Name  \
1240  1201-1400    Not Ranked  Islamic Azad University   

               Country/Territory Region Size Focus Research Status  AR SCORE  \
1240  Iran (Islamic Republic of)   Asia  NaN    CO      NaN    NaN       6.1   

      ... ISR RANK  ISD SCORE ISD RANK  IRN SCORE IRN RANK  EO SCORE EO RANK  \
1240  ...     801+        1.1     801+       96.3       55      35.7     482   

      SUS SCORE SUS RANK  Overall SCORE  
1240       29.4     801+            NaN  

[1 rows x 30 columns]

In [18]:
# Imputing 'Size'

df_2026_ranking['Size'].fillna('XL', inplace=True)

C:\Users\andd0\AppData\Local\Temp\ipykernel_23700\746180564.py:3: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df_2026_ranking['Size'].fillna('XL', inplace=True)


In [19]:
# Imputing 'Research'

df_2026_ranking['Research'].fillna('N/A', inplace=True)

C:\Users\andd0\AppData\Local\Temp\ipykernel_23700\175542628.py:3: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df_2026_ranking['Research'].fillna('N/A', inplace=True)


In [20]:
# Checking results

row_imputed = df_2026_ranking[df_2026_ranking['Institution Name'] == 'Islamic Azad University']

In [21]:
row_imputed

2026 Rank Previous Rank         Institution Name  \
1240  1201-1400    Not Ranked  Islamic Azad University   

               Country/Territory Region Size Focus Research Status  AR SCORE  \
1240  Iran (Islamic Republic of)   Asia   XL    CO      N/A    NaN       6.1   

      ... ISR RANK  ISD SCORE ISD RANK  IRN SCORE IRN RANK  EO SCORE EO RANK  \
1240  ...     801+        1.1     801+       96.3       55      35.7     482   

      SUS SCORE SUS RANK  Overall SCORE  
1240       29.4     801+            NaN  

[1 rows x 30 columns]

##### **Columns** *Status*

##### **Decision:** An online search can help find out the status of each one of the 47 universities. Sources: Wikipedia and Google AI

##### **Justification:** While Wikipedia and the AI are not sources one would resort to for more complex/sensitive info, the status of a university is something these sources can provide.

In [22]:
# Finding missing values under 'Status'

df_missing_status = df_2026_ranking[df_2026_ranking['Status'].isnull() == True]

In [23]:
df_missing_status

2026 Rank Previous Rank  \
81           82    Not Ranked   
261         262           236   
309         310    Not Ranked   
382         381           355   
487         487           380   
519         519           440   
529         530    Not Ranked   
572         571           489   
582         582           547   
784     781-790    Not Ranked   
799     791-800       691-700   
846     801-850    Not Ranked   
851     801-850       781-790   
852     851-900       771-780   
857     851-900    Not Ranked   
897     851-900       801-850   
919     901-950    Not Ranked   
922     901-950       901-950   
933     901-950       851-900   
934     901-950    Not Ranked   
985    951-1000      951-1000   
995    951-1000    Not Ranked   
1019  1001-1200     1201-1400   
1115  1001-1200      951-1000   
1136  1001-1200     1001-1200   
1138  1001-1200    Not Ranked   
1142  1001-1200      951-1000   
1154  1001-1200     1001-1200   
1157  1001-1200     1001-1200   
1159  1001-1200     1001-1200   
1162  1001-1200     1001-1200   
1181  1001-1200    Not Ranked   
1201  1201-1400    Not Ranked   
1240  1201-1400    Not Ranked   
1246  1201-1400     1201-1400   
1253  1201-1400     1201-1400   
1267  1201-1400    Not Ranked   
1300  1201-1400     1201-1400   
1312  1201-1400    Not Ranked   
1335  1201-1400     1201-1400   
1357  1201-1400     1201-1400   
1388  1201-1400     1201-1400   
1433      1401+     1201-1400   
1449      1401+    Not Ranked   
1452      1401+     1201-1400   
1463      1401+    Not Ranked   
1489      1401+    Not Ranked   

                                       Institution Name  \
81                                  Adelaide University   
261                              University of Aberdeen   
309              City St George’s, University of London   
382                            University of Luxembourg   
487        Technische Universität Bergakademie Freiberg   
519       Singapore University of Technology and Design   
529               The Education University of Hong Kong   
572                    Isfahan University of Technology   
582                               University of Iceland   
784                   Hong Kong Metropolitan University   
799         Norwegian University of Life Sciences (UMB)   
846                                 University of Namur   
851        Zurich University of Applied Sciences (ZHAW)   
852                              Addis Ababa University   
857                     Azerbaijan Technical University   
897                             University of Stavanger   
919                        Macao Polytechnic University   
922                       Osaka Metropolitan University   
933                           Universidad de Valladolid   
934                       Universidad Europea de Madrid   
985                     Universidad de Córdoba - España   
995                             University of Cyberjaya   
1019                                     CEU University   
1115                                Universidad de León   
1136                              Universitat de Lleida   
1138           Université de Bretagne Occidentale (UBO)   
1142                     Université Sorbonne Paris Nord   
1154                               University of Deusto   
1157                               University of Ibadan   
1159                             University of Ioannina   
1162                                University of Lagos   
1181                          University of the Algarve   
1201                     Ahmadu Bello University, Zaria   
1240                            Islamic Azad University   
1246                           Jahangirnagar University   
1253  Kwame Nkrumah University of Science and Techno...   
1267                        New Mexico State University   
1300              Technische Universität Kaiserslautern   
1312    UCAM Universidad Católica San Antonio de Murcia   
1335            Universidad Nacional de Ingeniería Peru   
13

##### I will create a dictionary composed by the universities with empty values under 'Status' and the online findings:

In [24]:
status_dict = {
    'Adelaide University': 'N/A',
	'University of Aberdeen': 'Public',
	'City St George’s, University of London': 'Public',
	'University of Luxembourg': 'Public',
	'Technische Universität Bergakademie Freiberg': 'Public',
	'Singapore University of Technology and Design': 'Public',
	'The Education University of Hong Kong': 'Public',
	'Isfahan University of Technology': 'Public',
	'University of Iceland': 'Public',
	'Hong Kong Metropolitan University': 'Public',
	'Norwegian University of Life Sciences (UMB)': 'Public',
	'University of Namur': 'Public',
	'Zurich University of Applied Sciences (ZHAW)': 'Public',
	'Addis Ababa University': 'Public',
	'Azerbaijan Technical University': 'Public',
	'University of Stavanger': 'Public',
	'Macao Polytechnic University': 'N/A',
	'Osaka Metropolitan University': 'Public',
	'Universidad de Valladolid': 'Public',
	'Universidad Europea de Madrid': 'Private for Profit',
	'Universidad de Córdoba - España': 'Public',
	'University of Cyberjaya': 'Private for Profit',
	'CEU University': 'Private not for Profit',
	'Universidad de León': 'Public',
	'Universitat de Lleida': 'Public',
	'Université de Bretagne Occidentale (UBO)': 'Public',
	'Université Sorbonne Paris Nord': 'Public',
	'University of Deusto': 'Private not for Profit',
	'University of Ibadan': 'Public',
	'University of Ioannina': 'Public',	
	'University of Lagos': 'Public',
	'University of the Algarve': 'Public',
	'Ahmadu Bello University, Zaria': 'Public',
	'Islamic Azad University': 'Private for Profit',
	'Jahangirnagar University': 'Public',
	'Kwame Nkrumah University of Science and Technology': 'Public',
	'New Mexico State University': 'Public',
	'Technische Universität Kaiserslautern': 'Public',
	'UCAM Universidad Católica San Antonio de Murcia':'Private not for Profit',
	'Universidad Nacional de Ingeniería Peru': 'Public',
	'Université de Limoges': 'Public',
	'University of Rajshahi': 'Public',
	'Khulna University': 'Public',
	'Rajshahi University of Engineering and Technology': 'Public',
	'San Francisco State University': 'Public',
	'Tongmyong University': 'Public',
	'University of Hawaii at Hilo': 'Public',
}

##### This mapping function will assign the findings to the corresponding universities

In [25]:
df_2026_ranking['Status'] = df_2026_ranking['Status'].fillna(df_2026_ranking['Institution Name'].map(status_dict))

In [26]:
# Let's check some universities to see if the change was successful

row_status = df_2026_ranking[df_2026_ranking['Institution Name'].isin([
    'Adelaide University',
    'University of Aberdeen',
    'City St George’s, University of London',
    'University of Luxembourg'
])]

In [27]:
row_status

2026 Rank Previous Rank                        Institution Name  \
81         82    Not Ranked                     Adelaide University   
261       262           236                  University of Aberdeen   
309       310    Not Ranked  City St George’s, University of London   
382       381           355                University of Luxembourg   

    Country/Territory   Region Size Focus Research  Status  AR SCORE  ...  \
81          Australia  Oceania   XL    CO       VH     N/A      78.3  ...   
261    United Kingdom   Europe    M    FC       VH  Public      40.1  ...   
309    United Kingdom   Europe    L    FO       VH  Public      26.9  ...   
382        Luxembourg   Europe    M    FO       VH  Public      13.2  ...   

    ISR RANK  ISD SCORE ISD RANK  IRN SCORE IRN RANK  EO SCORE EO RANK  \
81        96       72.5      251       96.7       46      70.4     207   
261      185       88.3      155       89.7      176      30.0     549   
309       90       98.0       77       77.2      402      65.8     238   
382       33      100.0       21       68.1      566      58.8     277   

     SUS SCORE SUS RANK  Overall SCORE  
81        86.1      98=           72.6  
261       87.8      83=           49.3  
309       62.3     452=           44.5  
382       42.4     801+           39.4  

[4 rows x 30 columns]

#### **Missing values in Ranks and Scores**

##### The IFR, ISR, IRN, and SUS ranks/scores are present in both rankings: 2025 and 2026. However, they don't use the same measuring methods, as the [2025](https://www.kaggle.com/datasets/melissamonfared/qs-world-university-rankings-2025?utm_source=chatgpt.com) one uses scores and the [other](https://www.kaggle.com/datasets/akashbommidi/2026-qs-world-university-rankings) ratios/percentages.

##### Regarding the ISD ranking/score (diversity among international students), it looks like is present in the 2026 ranking only - at least, it is not clear what its equivalent is in the 2025 ranking.

##### For these reasons, all empty values under these ranks columns will be labelled as 'Not Ranked' and the score columns will be replaced with NaN; thus, the columns' datatype (float) will not change to string.

#### **IFR**

In [28]:
df_ifr_rank_missing = df_2026_ranking [df_2026_ranking['IFR RANK'].isnull() == True]

In [29]:
df_ifr_rank_missing.shape

(87, 30)

In [30]:
df_ifr_score_missing = df_2026_ranking [df_2026_ranking['IFR SCORE'].isnull() == True]

In [31]:
df_ifr_score_missing.shape

(87, 30)

##### It matches the number of missing values under *4) Completeness* = 87. I proceed with the imputing.

In [32]:
df_2026_ranking['IFR RANK'].fillna('Not Ranked', inplace = True)

C:\Users\andd0\AppData\Local\Temp\ipykernel_23700\4224818905.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df_2026_ranking['IFR RANK'].fillna('Not Ranked', inplace = True)


In [33]:
df_2026_ranking['IFR SCORE'].fillna(np.nan, inplace = True)

C:\Users\andd0\AppData\Local\Temp\ipykernel_23700\2801279544.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df_2026_ranking['IFR SCORE'].fillna(np.nan, inplace = True)


In [34]:
# Checking results

df_2026_ranking[df_2026_ranking['IFR RANK'] == 'Not Ranked'][['Institution Name', 'IFR RANK', 'IFR SCORE']]

Institution Name    IFR RANK  IFR SCORE
215   Indian Institute of Technology Kharagpur (IITKGP)  Not Ranked        NaN
355             National Technical University of Athens  Not Ranked        NaN
361    University of Chinese Academy of Sciences (UCAS)  Not Ranked        NaN
464                                     Anna University  Not Ranked        NaN
519       Singapore University of Technology and Design  Not Ranked        NaN
...                                                 ...         ...        ...
1479                       Universidad Nacional del Sur  Not Ranked        NaN
1482        Universidade Federal do Rio Grande Do Norte  Not Ranked        NaN
1491                             University of Moratuwa  Not Ranked        NaN
1493                  University of Sri Jayewardenepura  Not Ranked        NaN
1496                         University of West Bohemia  Not Ranked        NaN

[87 rows x 3 columns]

#### **ISR**

In [35]:
df_isr_rank_missing = df_2026_ranking [df_2026_ranking['ISR RANK'].isnull() == True]

In [36]:
df_isr_rank_missing.shape

(37, 30)

In [37]:
df_isr_score_missing = df_2026_ranking [df_2026_ranking['ISR SCORE'].isnull() == True]

In [38]:
df_isr_score_missing.shape

(37, 30)

##### It matches the number of missing values under *4) Completeness* = 37. I proceed with the imputing.

In [39]:
df_2026_ranking['ISR RANK'].fillna('Not Ranked', inplace = True)

C:\Users\andd0\AppData\Local\Temp\ipykernel_23700\648179726.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df_2026_ranking['ISR RANK'].fillna('Not Ranked', inplace = True)


In [40]:
df_2026_ranking['ISR SCORE'].fillna(np.nan, inplace = True)

C:\Users\andd0\AppData\Local\Temp\ipykernel_23700\3384610585.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df_2026_ranking['ISR SCORE'].fillna(np.nan, inplace = True)


In [41]:
# Checking results

df_2026_ranking[df_2026_ranking['ISR RANK'] == 'Not Ranked'][['Institution Name', 'ISR RANK', 'ISR SCORE']]

Institution Name    ISR RANK  ISR SCORE
519       Singapore University of Technology and Design  Not Ranked        NaN
653           China University of Mining and Technology  Not Ranked        NaN
706        Nanjing University of Science and Technology  Not Ranked        NaN
724   Pakistan Institute of Engineering and Applied ...  Not Ranked        NaN
762   Bangladesh University of Engineering and Techn...  Not Ranked        NaN
788                            Universidad de Antioquia  Not Ranked        NaN
799         Norwegian University of Life Sciences (UMB)  Not Ranked        NaN
852                              Addis Ababa University  Not Ranked        NaN
871                                 Jiangnan University  Not Ranked        NaN
888                         Université de Franche-Comté  Not Ranked        NaN
897                             University of Stavanger  Not Ranked        NaN
985                     Universidad de Córdoba - España  Not Ranked        NaN
990            Universidad Nacional Mayor de San Marcos  Not Ranked        NaN
1032                      Georgian Technical University  Not Ranked        NaN
1038                                   Hohai University  Not Ranked        NaN
1047                   Kyrgyz Russian Slavic University  Not Ranked        NaN
1138           Université de Bretagne Occidentale (UBO)  Not Ranked        NaN
1142                     Université Sorbonne Paris Nord  Not Ranked        NaN
1147                  University of Alabama, Birmingham  Not Ranked        NaN
1157                               University of Ibadan  Not Ranked        NaN
1162                                University of Lagos  Not Ranked        NaN
1168                              University of Nairobi  Not Ranked        NaN
1216            California Polytechnic State University  Not Ranked        NaN
1246                           Jahangirnagar University  Not Ranked        NaN
1253  Kwame Nkrumah University of Science and Techno...  Not Ranked        NaN
1300              Technische Universität Kaiserslautern  Not Ranked        NaN
1336                   Universidad Nacional de San Luis  Not Ranked        NaN
1337  Universidad Nacional del Centro de la Provinci...  Not Ranked        NaN
1357                              Université de Limoges  Not Ranked        NaN
1381                             University of Novi Sad  Not Ranked        NaN
1417  Chittagong University of Engineering and Techn...  Not Ranked        NaN
1433                                  Khulna University  Not Ranked        NaN
1434    Khulna University of Engineering and Technology  Not Ranked        NaN
1449  Rajshahi University of Engineering and Technology  Not Ranked        NaN
1462      The Islamia University of Bahawalpur,Pakistan  Not Ranked        NaN
1471                            Universidad de Tarapaca  Not Ranked        NaN
1479                       Universidad Nacional del Sur  Not Ranked        NaN

#### **IRN**

In [42]:
df_irn_rank_missing = df_2026_ranking [df_2026_ranking['IRN RANK'].isnull() == True]

In [43]:
df_irn_rank_missing.shape

(2, 30)

In [44]:
df_irn_score_missing = df_2026_ranking [df_2026_ranking['IRN SCORE'].isnull() == True]

In [45]:
df_irn_score_missing.shape

(2, 30)

##### It matches the number of missing values under *4) Completeness* = 2. I proceed with the imputing.

In [46]:
df_2026_ranking['IRN RANK'].fillna('Not Ranked', inplace = True)

C:\Users\andd0\AppData\Local\Temp\ipykernel_23700\456051274.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df_2026_ranking['IRN RANK'].fillna('Not Ranked', inplace = True)


In [47]:
df_2026_ranking['IRN SCORE'].fillna(np.nan, inplace = True)

C:\Users\andd0\AppData\Local\Temp\ipykernel_23700\2800520121.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df_2026_ranking['IRN SCORE'].fillna(np.nan, inplace = True)


In [48]:
# Checking results

df_2026_ranking[df_2026_ranking['IRN RANK'] == 'Not Ranked'][['Institution Name', 'IRN RANK', 'IRN SCORE']]

Institution Name    IRN RANK  IRN SCORE
929  Universidad Católica Andrés Bello - UCAB  Not Ranked        NaN
984                   Universidad de Belgrano  Not Ranked        NaN

#### **ISD**

In [49]:
df_isd_rank_missing = df_2026_ranking [df_2026_ranking['ISD RANK'].isnull() == True]

In [50]:
df_isd_rank_missing.shape

(37, 30)

In [51]:
df_isd_score_missing = df_2026_ranking [df_2026_ranking['ISD SCORE'].isnull() == True]

In [52]:
df_isd_score_missing.shape

(37, 30)

##### It matches the number of missing values under *4) Completeness* = 37. I proceed with the imputing.

In [53]:
df_2026_ranking['ISD RANK'].fillna('Not Ranked', inplace = True)

C:\Users\andd0\AppData\Local\Temp\ipykernel_23700\367821517.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df_2026_ranking['ISD RANK'].fillna('Not Ranked', inplace = True)


In [54]:
df_2026_ranking['ISD SCORE'].fillna(np.nan, inplace = True)

C:\Users\andd0\AppData\Local\Temp\ipykernel_23700\297048921.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df_2026_ranking['ISD SCORE'].fillna(np.nan, inplace = True)


In [55]:
# Checking results

df_2026_ranking[df_2026_ranking['ISD RANK'] == 'Not Ranked'][['Institution Name', 'ISD RANK', 'ISD SCORE']]

Institution Name    ISD RANK  ISD SCORE
519       Singapore University of Technology and Design  Not Ranked        NaN
653           China University of Mining and Technology  Not Ranked        NaN
706        Nanjing University of Science and Technology  Not Ranked        NaN
724   Pakistan Institute of Engineering and Applied ...  Not Ranked        NaN
762   Bangladesh University of Engineering and Techn...  Not Ranked        NaN
788                            Universidad de Antioquia  Not Ranked        NaN
799         Norwegian University of Life Sciences (UMB)  Not Ranked        NaN
852                              Addis Ababa University  Not Ranked        NaN
871                                 Jiangnan University  Not Ranked        NaN
888                         Université de Franche-Comté  Not Ranked        NaN
897                             University of Stavanger  Not Ranked        NaN
985                     Universidad de Córdoba - España  Not Ranked        NaN
990            Universidad Nacional Mayor de San Marcos  Not Ranked        NaN
1032                      Georgian Technical University  Not Ranked        NaN
1038                                   Hohai University  Not Ranked        NaN
1047                   Kyrgyz Russian Slavic University  Not Ranked        NaN
1138           Université de Bretagne Occidentale (UBO)  Not Ranked        NaN
1142                     Université Sorbonne Paris Nord  Not Ranked        NaN
1147                  University of Alabama, Birmingham  Not Ranked        NaN
1157                               University of Ibadan  Not Ranked        NaN
1162                                University of Lagos  Not Ranked        NaN
1168                              University of Nairobi  Not Ranked        NaN
1216            California Polytechnic State University  Not Ranked        NaN
1246                           Jahangirnagar University  Not Ranked        NaN
1253  Kwame Nkrumah University of Science and Techno...  Not Ranked        NaN
1300              Technische Universität Kaiserslautern  Not Ranked        NaN
1336                   Universidad Nacional de San Luis  Not Ranked        NaN
1337  Universidad Nacional del Centro de la Provinci...  Not Ranked        NaN
1357                              Université de Limoges  Not Ranked        NaN
1381                             University of Novi Sad  Not Ranked        NaN
1417  Chittagong University of Engineering and Techn...  Not Ranked        NaN
1433                                  Khulna University  Not Ranked        NaN
1434    Khulna University of Engineering and Technology  Not Ranked        NaN
1449  Rajshahi University of Engineering and Technology  Not Ranked        NaN
1462      The Islamia University of Bahawalpur,Pakistan  Not Ranked        NaN
1471                            Universidad de Tarapaca  Not Ranked        NaN
1479                       Universidad Nacional del Sur  Not Ranked        NaN

#### **SUS**

In [56]:
df_sus_rank_missing = df_2026_ranking [df_2026_ranking['SUS RANK'].isnull() == True]

In [57]:
df_sus_rank_missing.shape

(24, 30)

In [58]:
df_sus_score_missing = df_2026_ranking [df_2026_ranking['SUS SCORE'].isnull() == True]

In [59]:
df_sus_score_missing.shape

(24, 30)

##### It matches the number of missing values under *4) Completeness* = 24. I proceed with the imputing.

In [60]:
df_2026_ranking['SUS RANK'].fillna('Not Ranked', inplace = True)

C:\Users\andd0\AppData\Local\Temp\ipykernel_23700\1441189629.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df_2026_ranking['SUS RANK'].fillna('Not Ranked', inplace = True)


In [61]:
df_2026_ranking['SUS SCORE'].fillna(np.nan, inplace = True)

C:\Users\andd0\AppData\Local\Temp\ipykernel_23700\2308821208.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df_2026_ranking['SUS SCORE'].fillna(np.nan, inplace = True)


In [62]:
# Checking results

df_2026_ranking[df_2026_ranking['SUS RANK'] == 'Not Ranked'][['Institution Name', 'SUS RANK', 'SUS SCORE']]

Institution Name    SUS RANK  SUS SCORE
361    University of Chinese Academy of Sciences (UCAS)  Not Ranked        NaN
446                         Belarusian State University  Not Ranked        NaN
778                              Universidad de Palermo  Not Ranked        NaN
859            Belarusian National Technical University  Not Ranked        NaN
929            Universidad Católica Andrés Bello - UCAB  Not Ranked        NaN
984                             Universidad de Belgrano  Not Ranked        NaN
995                             University of Cyberjaya  Not Ranked        NaN
1010    Asfendiyarov Kazakh National Medical University  Not Ranked        NaN
1033  German University of Technology in Oman  (GUTECH)  Not Ranked        NaN
1040                      Islamic University of Lebanon  Not Ranked        NaN
1047                   Kyrgyz Russian Slavic University  Not Ranked        NaN
1097                           Sumgait State University  Not Ranked        NaN
1110          Universidad Argentina de la Empresa -UADE  Not Ranked        NaN
1117                          Universidad de San Andrés  Not Ranked        NaN
1191                                Van Lang University  Not Ranked        NaN
1200                                     ADA University  Not Ranked        NaN
1230                               Galgotias University  Not Ranked        NaN
1249  Kazakh Ablai khan University of International ...  Not Ranked        NaN
1255                         Kyrgyz National University  Not Ranked        NaN
1300              Technische Universität Kaiserslautern  Not Ranked        NaN
1412  Belarusian State University of Informatics and...  Not Ranked        NaN
1430                                  Jadara University  Not Ranked        NaN
1475  Universidad Latinoamericana de Ciencia y Tecno...  Not Ranked        NaN
1476                          Universidad Metropolitana  Not Ranked        NaN

#### **Accuracy and consistency checks after 1st cleaning**

In [63]:
# Rows and columns

df_2026_ranking.shape

(1501, 30)

In [64]:
# Missing values

df_2026_ranking.isnull().sum()

2026 Rank              0
Previous Rank          0
Institution Name       0
Country/Territory      0
Region                 0
Size                   0
Focus                  0
Research               0
Status                 0
AR SCORE               0
AR RANK                0
ER SCORE               0
ER RANK                0
FSR SCORE              0
FSR RANK               0
CPF SCORE              0
CPF RANK               0
IFR SCORE             87
IFR RANK               0
ISR SCORE             37
ISR RANK               0
ISD SCORE             37
ISD RANK               0
IRN SCORE              2
IRN RANK               0
EO SCORE               0
EO RANK                0
SUS SCORE             24
SUS RANK               0
Overall SCORE        798
dtype: int64

In [65]:
# Data types

df_2026_ranking.dtypes

2026 Rank             object
Previous Rank         object
Institution Name      object
Country/Territory     object
Region                object
Size                  object
Focus                 object
Research              object
Status                object
AR SCORE             float64
AR RANK               object
ER SCORE             float64
ER RANK               object
FSR SCORE            float64
FSR RANK              object
CPF SCORE            float64
CPF RANK              object
IFR SCORE            float64
IFR RANK              object
ISR SCORE            float64
ISR RANK              object
ISD SCORE            float64
ISD RANK              object
IRN SCORE            float64
IRN RANK              object
EO SCORE             float64
EO RANK               object
SUS SCORE            float64
SUS RANK              object
Overall SCORE        float64
dtype: object

##### **Notes:** 

##### - The numbers different from 0 in front of the columns (first batch below 'shape') correspond to NaN and not empty cells.

##### - The data types remain as needed.

In [66]:
# Let's check descriptive stats after the cleaning:

df_2026_ranking.describe()

AR SCORE     ER SCORE    FSR SCORE    CPF SCORE    IFR SCORE  \
count  1501.000000  1501.000000  1501.000000  1501.000000  1414.000000   
mean     25.785943    26.944237    33.950433    30.425516    36.305658   
std      24.500905    25.504494    28.440071    29.679882    35.252024   
min       1.000000     1.000000     1.000000     1.000000     1.000000   
25%       8.800000     8.500000    10.800000     6.000000     6.600000   
50%      16.000000    16.500000    23.500000    18.000000    20.100000   
75%      32.700000    37.500000    50.500000    49.700000    66.100000   
max     100.000000   100.000000   100.000000   100.000000   100.000000   

        ISR SCORE    ISD SCORE    IRN SCORE     EO SCORE    SUS SCORE  \
count  1464.00000  1464.000000  1499.000000  1501.000000  1477.000000   
mean     33.32541    34.526981    53.356905    29.989674    51.254367   
std      32.75066    31.108600    28.920632    29.197573    21.266331   
min       1.00000     1.000000     1.000000     1.000000     3.000000   
25%       5.90000     8.675000    27.500000     6.200000    35.700000   
50%      19.40000    21.700000    55.800000    17.900000    48.700000   
75%      56.70000    55.650000    78.500000    46.000000    66.500000   
max     100.00000   100.000000   100.000000   100.000000   100.000000   

       Overall SCORE  
count     703.000000  
mean       46.756046  
std        18.842125  
min        25.100000  
25%        30.900000  
50%        41.500000  
75%        58.900000  
max       100.000000

In [67]:
# Mode

df_2026_ranking.mode().iloc[0]

2026 Rank                           1201-1400
Previous Rank                       1001-1200
Institution Name               ADA University
Country/Territory    United States of America
Region                                   Asia
Size                                        L
Focus                                      FC
Research                                   VH
Status                                 Public
AR SCORE                                  6.4
AR RANK                                  701+
ER SCORE                                  8.2
ER RANK                                  701+
FSR SCORE                               100.0
FSR RANK                                 801+
CPF SCORE                                 1.4
CPF RANK                                 801+
IFR SCORE                               100.0
IFR RANK                                 801+
ISR SCORE                               100.0
ISR RANK                                 801+
ISD SCORE                         

#### **Exporting**

In [68]:
df_2026_ranking.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'Advanced Analytics and Dashboard Design (I).csv'))